src/node/interfaces.cpp

```diff
@@ -819,18 +819,22 @@ class ChainImpl : public Chain
     {
         std::optional<interfaces::SettingsAction> action;
         args().LockSettings([&](common::Settings& settings) {
-            auto* ptr_value = common::FindKey(settings.rw_settings, name);
-            // Create value if it doesn't exist
-            auto& value = ptr_value ? *ptr_value : settings.rw_settings[name];
-            action = update_settings_func(value);
+            if (auto* value = common::FindKey(settings.rw_settings, name)) {
+                action = update_settings_func(*value);
+                if (value->isNull()) settings.rw_settings.erase(name);
+            } else {
+                UniValue new_value;
+                action = update_settings_func(new_value);
+                if (!new_value.isNull()) settings.rw_settings[name] = std::move(new_value);
+            }
         });
         if (!action) return false;
         // Now dump value to disk if requested
         return *action == interfaces::SettingsAction::SKIP_WRITE || args().WriteSettingsFile();
     }
     bool overwriteRwSetting(const std::string& name, common::SettingsValue value, interfaces::SettingsAction action) override
     {
-        if (value.isNull()) return deleteRwSettings(name, action);
+        if (value.isNull()) return false; // disallow overwriting with null values.
```

Comment:
In commit "chain: disallow overwriting settings with null values" (4627174d80838e1c6848f9dff47a2ca4a9a35091)

Would be good to just delete this line entirely and make overwriteRwSetting just call updateRwSetting without making a special case for null. Setting a null setting should just delete a setting consistently and not have different behavior for different functions.

(Sorry I didn't notice this line before and mistakenly thought overwriteRwSetting would allow writing null values to the settings file, so my earlier comments might have been confusing.)

Reply:
> Setting a null setting should just delete a setting consistently and not have different behavior for different functions.

I would rather fail when a null setting is provided to any of the update or overwrite functions. I believe the intent to delete should be explicit, at least within this settings layer. That's why the `deleteRwSettings` function was introduced. It seems inconsistent to have two different methods for deleting a value - one embedded within the update function and another explicit.

Reply:
So we both agree it is bad for overwrite function to reject null values and update function to accept them, but I would prefer both functions to accept them and you want both functions to reject them.

Here are the reason I think it is better to accept them:

- I think it is generally bad practice to accept nullable values and then handle them with runtime errors. If you want to write API's that don't accept null values I think you should use non-nullable types.

- I think in this specific case, there is no reason these APIs should not accept null values. Ihe purpose of the update function is to be able to atomically read and update settings. But there is no function to atomically read and delete settings and this should be a completely valid use-case for situations where you want to safely reset settings without clobbering concurrently set values. We can easily support it by just not going out of the way to add unnecessary restrictions.

I am fine with existence of deleteRwSetting as a more convenient & readable wrapper around overwriteRwSetting and updateRwSetting, the same way overwriteRwSetting is a convenient and readable wrapper around updateRwSetting. But existence of deleteRwSetting should not be an excuse for crippling the behavior of the other functions and making them less safe to call. Ability to atomically read and delete settings is useful, and trying to deal with null values at runtime rather than compile time is a bad practice that should be a last not first resort.

Reply:
There are two suggestions here:

1. **@furszy suggested to fail when the value is null**: I agree with not checking for null values, for the reasons outlined above.
   
2. **@ryanofsky suggested removing the null value check**: This approach would overwrite the previous settings with a null value.
   
> Setting a null value should consistently delete the setting, rather than having different behavior in different functions.

However, if we don’t delete the key-value pair, the setting still exists in the map with a null value, which does not truly remove the setting, right?

Currently, the code calls `deleteRwSettings` when the `value` is null, which deletes the setting entirely. This makes the method redundant, as calling `deleteRwSettings` achieves the same result as passing a null `SettingsValue` to `overwriteRwSetting`.

I believe a better approach is to simply update the settings based on what is passed to `overwriteRwSetting`.
It should be the caller's responsibility to determine whether to pass a null value or not. 
If the caller intends to delete the setting, they should explicitly call `deleteRwSettings`.

Reply:
> I think the only thing that "bothers" me a bit about the nullable class is the different meanings it could have, which require additional clarifications at the API level. 

Yeah this is real concern, and I think a common thing that happens with null values, true/false values, empty strings, and special integer values. If you are sloppy with these values you will run into problems. If you are thoughtful you can come up with very clean interfaces, but it is hard to be careful. You can avoid problems with ambiguous values by using enums, std::variant, not_null implementations, and custom types, but these can be inconvenient or verbose. We're just trying to pick a tradeoff that is convenient and doesn't cause big surprises.